### ISSUE TITLE
AddEFConfiguration does not exist in Microsoft.Extensions.Configuration.Memory

### ISSUE THREAD

**Issue by Micteu** [dotnet/AspNetCore.Docs#21482]
In the section https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/configuration?view=aspnetcore-5.0#custom-configuration-provider-with-ef-core, the directions say to add 

```
builder.Configuration.AddEFConfiguration(
    options => options.UseInMemoryDatabase("InMemoryDb"));
```

However, I get this error:

'WebAssemblyHostConfiguration' does not contain a definition for 'AddEFConfiguration' and no accessible extension method 'AddEFConfiguration' accepting a first argument of type 'WebAssemblyHostConfiguration' could be found (are you missing a using directive or an assembly reference?)

and `Microsoft.Extensions.Configuration.Memory` is an unused using. This is in a .NET 5 app, and `Microsoft.EntityFrameworkCore` and `Microsoft.EntityFrameworkCore.InMemory` are both version 5.0.2.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7367a651-2b27-0a57-e7d0-95686cc97f1c
* Version Independent ID: 8c769648-728d-72ce-ebf6-2a06e5076aa2
* Content: [ASP.NET Core Blazor configuration](https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/configuration?view=aspnetcore-5.0)
* Content Source: [aspnetcore/blazor/fundamentals/configuration.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/fundamentals/configuration.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Hello @Micteu ... This guidance is proably a point-in-time scenario for 3.x apps, when it was confirmed working when it was put up. I'll get back to you soon.

**Comment by guardrex** [comment]
@captainsafia ... This one goes way, **_way_** back. 👴 It was a thing back in 🏛️ *ancient Blazor times* 🏛️, and it was a working scenario here, hence the detailed coverage where you can see that I was using it in a test app at one point. There was the engineering repo shake-up in the meantime, and that might explain the loss of a direct reference to the API. It's also possible that the extension method was created in the namespace in the app locally but didn't make it into the coverage (but that feels unlikely to me). Either way, I can't find the API. Do you concur that we should completely drop this coverage at this point?

... and ping @JeremyLikness ... to perhaps recall this scenario as well and remark further on **_any_** known EF Core config provider specifically for Blazor ... or even a non-Blazor scenario that would work in a Blazor app ... and even if any such approach would be for Blazor Server only. 😅 ... *Lot's of asks there!* 😄

**Comment by JeremyLikness** [comment]
Hi,

I'm 99% confident that the extension `AddEFConfiguration` does _not_ come from the EF Core code base. As for Blazor extensions, what's provided out of the box are all designed for standard ASP.NET Core DI configuration so most extend service provider. For example, there are the `AddDbContext` and `AddDbContextFactory` extensions which both take an options lambda for configuration.

**Comment by guardrex** [comment]
Thanks @JeremyLikness ... that's what I thought. It **_was_** a thing on our side, but it was dropped at some point AFAICT.

I'll wait for @captainsafia to concur on dropping this section. Doing something like this would be a dev-created scenario using the WASM-EFCore topic at this point, and we won't provide a custom solution (probably).

**Comment by guardrex** [comment]
@JeremyLikness ... **_I found it!_** ... It was in one of my (perhaps 100 😵) sample apps here fully implemented. Indeed ... and idk exactly why ... the section isn't complete. I think churn caused this. We've re-organized the config docs perhaps **_three times_**, and this example lost some bits on one of those updates. *Sorry about that @Micteu!*

@JeremyLikness ...

* Is the following sane/supported?
* If so, is it ok for Blazor WASM, Blazor Server, or both?
* If it is ok for WASM, I suspect that we need to call out the WebAssembly sandbox restriction on making direct dB connections and say that the example works for an in-memory (or similar) dB scenarios (we can work out the exact lingo from a draft that I'll put on a PR). If we call that out, I'd like to cross-link to something or provide additional detail.
* If this isn't good to show for either hosting model, then we'll proceed with the PR that I just opened (#21490) and 🔪 this whole section from the topic.

Here's the full code ...

`EFConfigurationValue.cs`:

```csharp
public class EFConfigurationValue
{
    public string Id { get; set; }
    public string Value { get; set; }
}
```

`EFConfigurationContext.cs`:

```csharp
using Microsoft.EntityFrameworkCore;

public class EFConfigurationContext : DbContext
{
    public EFConfigurationContext(DbContextOptions options) : base(options)
    {
    }

    public DbSet<EFConfigurationValue> Values { get; set; }
}
```

`EFConfigurationProvider.cs`:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;

public class EFConfigurationProvider : ConfigurationProvider
{
    public EFConfigurationProvider(Action<DbContextOptionsBuilder> optionsAction)
    {
        OptionsAction = optionsAction;
    }

    Action<DbContextOptionsBuilder> OptionsAction { get; }

    public override void Load()
    {
        var builder = new DbContextOptionsBuilder<EFConfigurationContext>();

        OptionsAction(builder);

        using (var dbContext = new EFConfigurationContext(builder.Options))
        {
            dbContext.Database.EnsureCreated();

            Data = !dbContext.Values.Any()
                ? CreateAndSaveDefaultValues(dbContext)
                : dbContext.Values.ToDictionary(c => c.Id, c => c.Value);
        }
    }

    private static IDictionary<string, string> CreateAndSaveDefaultValues(
        EFConfigurationContext dbContext)
    {
        // Quotes (c)2005 Universal Pictures: Serenity
        // https://www.uphe.com/movies/serenity
        var configValues =
            new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                { "quote1", "I aim to misbehave." },
                { "quote2", "I swallowed a bug." },
                { "quote3", "You can't stop the signal, Mal." }
            };

        dbContext.Values.AddRange(configValues
            .Select(kvp => new EFConfigurationValue
            {
                Id = kvp.Key,
                Value = kvp.Value
            })
            .ToArray());

        dbContext.SaveChanges();

        return configValues;
    }
}
```

`EFConfigurationSource.cs`:

```csharp
using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;

public class EFConfigurationSource : IConfigurationSource
{
    private readonly Action<DbContextOptionsBuilder> _optionsAction;

    public EFConfigurationSource(Action<DbContextOptionsBuilder> optionsAction)
    {
        _optionsAction = optionsAction;
    }

    public IConfigurationProvider Build(IConfigurationBuilder builder)
    {
        return new EFConfigurationProvider(_optionsAction);
    }
}
```

`EFExtensions.cs`:

```csharp
using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;

public static class EFExtensions
{
    public static IConfigurationBuilder AddEFConfiguration(
        this IConfigurationBuilder builder,
        Action<DbContextOptionsBuilder> optionsAction)
    {
        return builder.Add(new EFConfigurationSource(optionsAction));
    }
}
```

`Program.cs`:

```csharp
builder.Configuration.AddEFConfiguration(
    options => options.UseInMemoryDatabase("InMemoryDb"));
```

`EFCoreConfig.razor`:

```razor
@page "/efcore-config"
@using Microsoft.Extensions.Configuration
@inject IConfiguration Configuration

<ul>
    <li>@Configuration["quote1"]</li>
    <li>@Configuration["quote2"]</li>
    <li>@Configuration["quote3"]</li>
</ul>
```

Output:

&bullet; I aim to misbehave.
&bullet; I swallowed a bug.
&bullet; You can't stop the signal, Mal.

**Comment by JeremyLikness** [comment]
@guardrex,

> * Is the following sane/supported?

I'm not familiar with the whole project so I'm not sure what this is trying to accomplish. What is the point of having key/value pairs on the `DbContext` vs. in some independent configuration entity? I'm sure the example explains this but I haven't had time to dig into details. Nothing insane or unsupported that I see.

> * If so, is it ok for Blazor WASM, Blazor Server, or both?

In-memory DB should be fine on server or WASM, with the caveats listed in [this document](https://docs.microsoft.com/ef/core/providers/in-memory/).

> * If it is ok for WASM, I suspect that we need to call out the WebAssembly sandbox restriction on making direct dB connections and say that the example works for an in-memory (or similar) dB scenarios (we can work out the exact lingo from a draft that I'll put on a PR). If we call that out, I'd like to cross-link to something or provide additional detail. I'm 👂 for more info on that.

WASM works with in-memory, which is not designed for production, and Azure Cosmos DB using an HTTP proxy and with security caveats (i.e. you should NOT save the connection string in the app, the only secure way to access it directly is to interact with an authenticated service that provides temporary [resource tokens](https://docs.microsoft.com/azure/cosmos-db/secure-access-to-data#resource-tokens-).

> * If this isn't good to show for either hosting model, then we'll proceed with the PR that I just opened (#21490) and 🔪💀 this whole section from the topic.

TL;DR; I think this is the best course of action to avoid confusion and heading down the wrong path.

Jeremy

**Comment by guardrex** [comment]
I think so, too ... 🔪💀. 

AFAIK, this was put here to merely show how to make a formal configuration provider via EF Core, and I think it was intended to be a scenario only for Blazor Server. We have good coverage for how to do that in a general way already in the ASP.NET Core doc on *Configuration*. If someone wants to make one of these, they can do so following the general guidance. If someone asked me about it specifically for Blazor Server, I would just point them to ... well ... **_here_** :smile: ... I mean now that I've pasted all of the code bits into a comment.

It does seem silly to do this client-side and in-memory for testing something that devs shouldn't be doing generally in production. We do say in the NOTE that ...

> Running EF Core in Blazor WebAssembly is beyond the scope of this article.

... which implies that it might be possible under some limited, edge case scenarios. When I get to that doc on an upcoming UE pass, we can take a look at how that's phrased and possibly flesh it out or flat out dissuade anyone from doing anything like that.

I'll just go ahead with my #21490 PR. Thanks for your help, @JeremyLikness! 🎷
